### PR TITLE
add OptimisticFilter extension

### PR DESCRIPTION
**Pull request by MaceWindu**
Requested here #4009

leverage use of optimistic lock with linq api

### CODE REVIEW

**Comment by jods4** [comment]
Bikeshedding: I'm not a fan of the name.
Before I looked at the code, it didn't cross my mind that it included a WHERE on the PK.

Do we have an API to look up objects by PK?
Given we have `Update`, `UpdateOptimistic` and `Delete`, `DeleteOptimistic`; it would make sense to name theses `Find` and `FindOptimistic` (assuming we have a `Find` that adds a WHERE by PK).

`Find` if the method doesn't exist yet, is also up for discussion, it's not a great name...

**Comment by jods4** [comment]
I think there are opportunities for refactoring here.

Every call to `MakeConcurrentFilter` (there are 3) is preceded by:
```csharp
var objType = typeof(T);
var ed      = dc.MappingSchema.GetEntityDescriptor(objType, dc.Options.ConnectionOptions.OnEntityDescriptorCreated);
```
Then `ed` is passed is last parameter of `MakeConcurrentFilter`.

-> This code should be in `MakeConcurrentFilter`, which should take the `dc` instead of `ed`.

Once that change is made, there's so little in Delete/Update methods that they can be rewritten as shorthands for what #4009 asked:
```csharp
public static int DeleteOptimistic<T>(this IQueryable<T> source, T obj)
  where T : class
  => source.FindOptimistic(obj).Delete();
```

(I also removed the null checks, but those will be done by `FindOptimistic`)

**Comment by MaceWindu** [comment]
both variants supported https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB.Tools/DataModel/Model/Entities/FindTypes.cs

**Comment by MaceWindu** [comment]
@jods4 so, what we are doing here (except optimization)? Should we change name?

**Comment by MaceWindu** [comment]
Key could be composite and I really don't want to invest too many time into this feature :)

**Comment by jods4** [comment]
@MaceWindu yes, I thought about that and figured: if your key is composite, just use the entity syntax or an explicit `Where`.

If you don't want to invest too much time, I don't think we need to implement everything right now.
We just need to be sure we're ok with naming and that it would be consistent if we expand the API later.

Personally, I think `WhereKeyOptimistic` is my favorite name so far.

The rest could be added later, and I think we can make `WhereKey<T>` work with both entities and values.
For a start, it would be 2 different overloads: `WhereKey<T>(this IQueryable<T> src, T entity)` and `WhereKey<T, K>(this IQueryable<T> src, K key)`.

File: Source/LinqToDB/Concurrency/ConcurrencyExtensions.cs [hunk 1]
```diff
@@ -241,9 +241,7 @@ public static int DeleteOptimistic<T>(this IQueryable<T> source, T obj)
 			if (source == null) throw new ArgumentNullException(nameof(source));
 			if (obj    == null) throw new ArgumentNullException(nameof(obj));
```
Comment (jods4) [inline Source/LinqToDB/Concurrency/ConcurrencyExtensions.cs:243]: @MaceWindu For simple methods that simply forward a call to another, which will then null-check the same parameters, don't you think it's ok to remove the checks?

**Review by jods4 — APPROVED** [review]

**Pull request merged**